main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -29,16 +29,21 @@ import scala.jdk.CollectionConverters.CollectionHasAsScala
 object Safety {
 
   /** Checks the safety and well-formedness of `root`. */
-  def run(root: Root)(implicit flix: Flix): (Root, List[SafetyError]) = flix.phaseNew("Safety") {
+  def run(root: Root, oldRoot: Root, changeSet: ChangeSet)(implicit flix: Flix): (Root, List[SafetyError]) = flix.phaseNew("Safety") {
     implicit val sctx: SharedContext = SharedContext.mk()
-    ParOps.parMap(root.traits.values.flatMap(_.sigs))(visitSig)
-    ParOps.parMap(root.defs.values)(visitDef)
+
+    val (staleDefs, freshDefs) = changeSet.partition(root.defs, oldRoot.defs)
+    val (staleSigs, freshSigs) = changeSet.partition(root.sigs, oldRoot.sigs)
+    val defs = freshDefs ++ ParOps.parMapValues(staleDefs)(visitDef)
+    val sigs = freshSigs ++ ParOps.parMapValues(staleSigs)(visitSig)
     ParOps.parMap(TypedAstOps.instanceDefsOf(root))(visitDef)
```

Comment:
In a new separate pr, add this function to change set:

```
   def partitionOnValues[K, V1 <: Sourceable, V2](newMap: Map[K, List[V1]], oldMap: Map[K, List[V2]]): (Map[K, List[V1]], Map[K, List[V2]]) 
```

and use it for `root.instances`.

Reply:
Oh, and then the maps should not be merged with ++, but with something that concatenates the two lists in the codomain.